### ISSUE TITLE
It should be safe to deallocate an RLMObject from any thread

### ISSUE THREAD

**Issue by RyanCopley** [realm/realm-cocoa#1893]
https://github.com/RyanCopley/Realm-AFNetworking-Crash

I'm not sure if this is a problem with how I'm doing things, Realm, or AFNetworking, but I have a proof of concept of the crash. When the Foo object deallocates due to a failed request, it is on a "different" thread. AFNetworking is setting the block to nil. The block is *never* called (to my knowledge, no breakpoints would catch).

I'm probably doing something naive, but this is very odd. I'm having to create a block that I strongly capture to keep the AFNetworking block alive... which I believe is creating a leak in my project. I've never had to watch my threading as much before Realm... :/

**Comment by RyanCopley** [comment]
Alternatively, 
```
-(void)auth {
    __weak typeof(self) weakSelf = self;
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    manager.responseSerializer = [AFHTTPResponseSerializer new];
    [manager GET:@"http://reddit.com/" parameters:nil success:^(AFHTTPRequestOperation *operation, id responseObject) {
        //Weakly referencing `self` doesn't work either.
        weakSelf.bar = [responseObject objectForKey:@"something"];
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        NSLog(@"Error: %@", error);
    }];
}
```

doesn't work either, as weakSelf becomes nil (because self deallocates)

**Comment by segiddins** [comment]
Running your example, changing `self` to be weak there fixes the issue for me.

**Comment by RyanCopley** [comment]
Look closer, `self` becomes nil, which is still not right

**Comment by segiddins** [comment]
Well, it's "right" in the sense that that is precisely what weak pointers do. There's an impedance mismatch between what you want to do and the approach you're taking -- your best bet is to refetch your objects every time you context switch onto a new queue.

**Comment by RyanCopley** [comment]
Yes, however, strongly capturing `self` in the block keeps the object alive, and AFNetworking will release it (eventually), stopping a retain cycle. I'm not suggesting weak references, it isn't what I am wanting to do, but referencing `self` strongly in the block should keep my object alive until my block deallocates.

**Comment by segiddins** [comment]
Except you’re not allowed to use persisted RLMObjects from a thread other than the one it was obtained on, and strongly referencing it in that block means you are violating Realm’s threading contract.

**Comment by RyanCopley** [comment]
Which I'm not, they're both on the main thread.

**Comment by segiddins** [comment]
Except AFNetworking obviously is retaining and releasing those blocks on a different thread (or at least is capable of doing so).

**Comment by timanglade** [comment]
@RyanCopley I sent you an email just now. Apologies from Realm again. Thanks for taking the time to file a bug + for providing a repro case. If this issue needs to be investigated any further @jpsim, @mrackwitz or @bdash can help.

**Comment by hermanolsson** [comment]
We are seeing the same crash since upgrading from 0.91.x -> 0.92.2 on a swift project using Realm-ObjC (via cocoapods) and no AFNetworking.

**Comment by jpsim** [comment]
Thanks for sharing your experiences regarding RLMObject deallocation.

As others have said, RLMObject's are only designed to be used from the thread by which they were created or obtained. This design has been the case since Realm was publicly available.

Only recently did we introduce (#1727 / v0.92.0) a check to ensure `[RLMObject dealloc]` was only ever called from the expected thread. This check is necessary to prevent accessors from being in an undesirable state.

Even though previous versions of Realm wouldn't exhibit any problems if an RLMObject was deallocated from a different thread in 99% of cases, the remaining 1% would yield incorrect results, and we consider this to be unacceptable. (those figures are a rough estimate). So if you care about the integrity of your data, downgrading Realm to avoid this issue is not a viable solution.

However, as the examples you have provided demonstrate, it is often difficult to ensure that 3rd party code won't deallocate an RLMObject on a different thread. Therefore, we'll be investigating ways to make deallocating RLMObjects on *any* thread succeed.

We consider this a high-priority item that we must do before a 1.0 release.

Until such a mechanism is in place, we recommend that you avoid deallocating an RLMObject from a different thread, which may sometimes involve doing non-obvious things such as weakly capturing `self` in a block as discussed here.

**Comment by RyanCopley** [comment]
Thank you @jpsim for looking into this, if I can be of any assistance please feel free to reach out.

**Comment by jpsim** [comment]
@tgoyne did some really good work to fix this issue in realm's core today. It still needs work, some optimization (maybe) and to be reviewed, but you should buy him a beer if you bump into him.

**Comment by hermanolsson** [comment]
Thank you @jpsim and @tgoyne ! After our discussion about this I had another look, and it was pretty clear what actually caused our issue. The biggest problem for us was the missing stack trace, and since my understanding of the issue was better after the talk I knew where to focus my efforts.

**Comment by RyanCopley** [comment]
@jpsim @tgoyne Any update on this issue? :) I may end up pulling all my hair out over this :p

**Comment by RyanCopley** [comment]
Ooh... for people using AFNetworking, 

```
    manager.completionQueue = dispatch_get_current_queue();
```

helps out a lot. There is a warning because `dispatch_get_current_queue()` is deprecated, but I'm not in the mood for caring about a deprecation warning right now

**Comment by mamaral** [comment]
It looks like I'm running into this issue as well, want to ensure its related. I'm using SDWebImage to download/cache remote images, and when images are loaded from the cache or downloaded I grab the image and save the NSData to my RLMObject. Every now and then SDWebImage will cancel a download, which triggers to following chunk of code:

```
cancelled:^{
   SDWebImageDownloader *sself = wself;
   if (!sself) return;
   dispatch_barrier_async(sself.barrierQueue, ^{
      [sself.URLCallbacks removeObjectForKey:url];
   });
}];
```

Following the stack track, removing that object from the array triggers the `dealloc` on `RLMObjectBase` and the subsequent `RLMCheckThread(_realm);` to fail.

Does this sound like its related to this issue?

**Comment by RyanCopley** [comment]
```yourSDWebImageDownloader.barrierQueue = dispatch_get_current_queue();```

**Comment by RyanCopley** [comment]
Yes it does.

**Comment by mamaral** [comment]
@RyanCopley I believe dispatch_get_main_queue() is the non-deprecated version of that, I'll try that out.

**Comment by RyanCopley** [comment]
@mamaral I believe I misread your code, but you can try that. `dispatch_get_main_queue()` is not an equivalent, however, but it may be in your case

**Comment by mamaral** [comment]
Indeed you're right they are not equivalent... I `misread yours`.

**Comment by mamaral** [comment]
Yeah I'm not sure what the solution is on this one - Every different method I've tried to integrate Realm with SDWebimage to save images downloaded as NSData on RLMObjects results in the threading issue.

**Comment by jpsim** [comment]
Status update: a pull request making it safe to deallocate an `RLMObject` and `RLMResults` from any thread was merged into realm's core db engine yesterday and a new version will be released as soon as possible. Thanks for your patience.

**Comment by RyanCopley** [comment]
That sounds great! Thank you so much for all your hard work

**Comment by mamaral** [comment]
Sweet, thanks guys!

**Comment by jpsim** [comment]
Fixed in #1984. Thanks for your patience. We'll be cutting a new release of Realm Objective-C and Realm Swift shortly.

**Comment by semireg** [comment]
We just ran across this bug and confirmed fix in v93.1. Thank you, @jpsim - Caylan
